packaging/macOS/freeze_program.py

```diff
@@ -0,0 +1,186 @@
+# Parsec Cloud (https://parsec.cloud) Copyright (c) AGPLv3 2016-2021 Scille SAS
+
+import os
+import sys
+import re
+from typing import Optional
+import shutil
+import argparse
+import subprocess
+from hashlib import sha256
+from pathlib import Path
+
+
+# Fully-qualified path for the executable should be used with subprocess to
+# avoid unreliability (especially when running from within a virtualenv)
+python = shutil.which("python")
+if not python:
+    raise SystemExit("Cannot find python command :(")
+
+
+BUILD_DIR = Path("build").resolve()
+
+TOOLS_VENV_DIR = BUILD_DIR / "tools_venv"
+DEFAULT_WHEEL_IT_DIR = BUILD_DIR / "wheel_it"
+
+PYTHON_VERSION = f"{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}"
+
+
+def run(cmd, **kwargs):
+    if isinstance(cmd, str):
+        cmd = cmd.split()
+    print(f">>> {' '.join(map(str, cmd))}")
+    ret = subprocess.run(cmd, check=True, **kwargs)
+    return ret
+
+
+def main(
+    program_source: Path,
+    include_parsec_ext: Optional[Path] = None,
+    wheel_it_dir: Optional[Path] = None,
+):
+    BUILD_DIR.mkdir(parents=True, exist_ok=True)
+
+    # Retrieve program version
+    global_dict = {}
+    if wheel_it_dir is not None:
+        candidates = list(wheel_it_dir.glob(f"parsec_cloud-*.whl"))
+        if len(candidates) == 0:
+            raise SystemExit(f"Parsec wheel not present in {wheel_it_dir}")
+        elif len(candidates) > 1:
+            raise SystemExit(f"Multiple possible Parsec wheels: {candidates}")
+        program_wheel = candidates[0]
+        match = re.match(r"parsec_cloud-([0-9][0-9\.+a-z]+)-.*.whl", program_wheel.name)
+        if not match:
+            raise SystemError(
+                f"Found Parsec wheel {program_wheel} but could not determine it version"
+            )
+        print(f"Parsec wheel is: {program_wheel}")
+        program_version = f"v{match.group(1)}"
+    else:
+        exec((program_source / "parsec/_version.py").read_text(), global_dict)
+        program_version = global_dict.get("__version__")
+    assert program_version.startswith("v")
+    program_version_without_v_prefix = program_version[1:]
+    print(f"### Detected Parsec version {program_version} ###")
+
+    # Bootstrap tools virtualenv
+    tools_python = TOOLS_VENV_DIR / "bin/python"
+    if not TOOLS_VENV_DIR.is_dir():
+        print("### Create tool virtualenv ###")
+        run(f"{ python } -m venv {TOOLS_VENV_DIR}")
```

Comment:
TODO => Cache Venv

Reply:
The ` if not TOOLS_VENV_DIR.is_dir()` is a way to reuse the virtualenv if it exists.
This is mostly done to ease the testing when modifying the script, otherwise this script is designed to be mostly a one shot run: typically when using it you want to start from scratch to be sure you're not packaging something outdated.

(On top of that pip is already caching the dependencies wheels)

Reply:
The comment was more for the CI

Reply:
I'd rather not use virtualenv cache (so only caching wheels from pip own cache) for packaging job to avoid misconfiguration or bug that could lead to a invalid generated release ;-)